File: 19-balanced-tree/solution.go
// Height-Balanced Binary Tree
// Time: O(n), Space: O(h) where h is height

package main

import (
	"fmt"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Solution struct{}

func (s *Solution) isBalanced(root *TreeNode) bool {
	return s.height(root) != -1
}

func (s *Solution) height(node *TreeNode) int {
	if node == nil {
		return 0
	}
	
	leftHeight := s.height(node.Left)
	if leftHeight == -1 {
		return -1
	}
	
	rightHeight := s.height(node.Right)
	if rightHeight == -1 {
		return -1
	}
	
	if int(math.Abs(float64(leftHeight-rightHeight))) > 1 {
		return -1
	}
	
	if leftHeight > rightHeight {
		return leftHeight + 1
	}
	return rightHeight + 1
}

// Test
func main() {
	sol := &Solution{}
	// Balanced tree
	root1 := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
	// Unbalanced tree
	root2 := &TreeNode{Val: 1, Left: &TreeNode{Val: 2, Left: &TreeNode{Val: 3}}}
	
	fmt.Println(sol.isBalanced(root1)) // true
	fmt.Println(sol.isBalanced(root2)) // false
}

File: 19-balanced-tree/solution.ts
// Height-Balanced Binary Tree
// Time: O(n), Space: O(h) where h is height

class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    
    constructor(val: number = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    isBalanced(root: TreeNode | null): boolean {
        const height = (node: TreeNode | null): number => {
            if (!node) return 0;
            
            const leftHeight = height(node.left);
            if (leftHeight === -1) return -1;
            
            const rightHeight = height(node.right);
            if (rightHeight === -1) return -1;
            
            if (Math.abs(leftHeight - rightHeight) > 1) return -1;
            
            return Math.max(leftHeight, rightHeight) + 1;
        };
        
        return height(root) !== -1;
    }
}

// Test
if (require.main === module) {
    const sol = new Solution();
    // Balanced tree
    const root1 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    // Unbalanced tree
    const root2 = new TreeNode(1, new TreeNode(2, new TreeNode(3)), null);
    
    console.log(sol.isBalanced(root1));  // true
    console.log(sol.isBalanced(root2));  // false
}

export { Solution, TreeNode };
